docs/design/node_taint_eviction.md

```diff
@@ -42,51 +48,42 @@ automaticReplacementOptions:
-
-The [update_status](https://github.com/FoundationDB/fdb-kubernetes-operator/blob/main/controllers/update_status.go) subreconciler will be extended to mark process groups running on a tainted node with the new condition.
-This can be achieved by iterating over all Pods and make requests to the node object, one example implementation could look like this:
-
-```go
-func checkProcessGroupsRunningOnTaintedNodes(pods []*corev1.Pod, status *fdbv1beta2.FoundationDBClusterStatus) {
-    if !cluster.NodeFailureDetectionEnabled() {
-        return
-    }
-
-    nodeTaintedMap := internal.GetNodes(pods) // This will generate a map containing the node name as key and a boolean value indicating if the node is "tainted"
-    for node, _ := range nodeMap {
-        // Check if node has a taint for the minimum duration in that case we will set the value to true.
-    }
-
-    podMap := internal.CreatePodMap(cluster, pods)
-    for _, processGroup := range status.ProcessGroups {
-        pod, podExists := podMap[processGroup.ProcessGroupID]
-        if !podExists {
-            continue
-        }
-
-        tainted, nodeFound := nodeTaintedMap[pod.Spec.NodeName]
-        if !nodeFound {
-            continue
-        }
-
-        processGroup.UpdateCondition(fdbv1beta2.RunningOnTaintedNode, tainted, status.ProcessGroups, processGroup.ProcessGroupID)
-    }
-}
+In the above configuration, it takes at least 3000 seconds (1200 + 1800) for the operator to replace a Pod on a Node with taint key `example.com/disconnected`, because it takes `durationInSeconds`, i.e., 1200 seconds, to mark the Pod to be replaced and it takes another `taintReplacementTimeSeconds`, i.e., 1800 seconds, to start replacing the Pod.
+
+**Requirement 3.** The feature lists the taint status for process groups. This allows users to have a holistic view of which Pods in a FDB cluster are running on tainted Nodes and being replaced.
+
+### Implementation
+We need to make the following four type of changes:
+
+**CRD change.** We need to extend the `automaticReplacementOptions` with the new fields `taints` and `taintReplacementTimeSeconds`
+
+**Detect Node taint and change Pod condition.**  We can piggy-back the existing `validateProcessGroup()` function in `update_status` reconciler to detect if a Pod is running on a tainted Node and mark the Pod's taint-related condition accordingly.
+
+A Pod's Node can be listed by
 ```
+node = &corev1.Node{}
+err := r.Get(ctx, client.ObjectKey{Name: pod.Spec.NodeName}, node)
+```
+
+**Status of Pod taint-related condition.** We will add two  [ProcessGroupConditionType](https://github.com/FoundationDB/fdb-kubernetes-operator/blob/main/api/v1beta2/foundationdbcluster_types.go#L673): `NodeTaintDetected`, which represents a Pod's Node is tainted but not long enough for operator to replace it, and `NodeTaintReplacing`, which represents a Pod whose Node has been tainted for longer-than-configured time and the operator should replace it.
```

Comment:
```suggestion
**Status of ProcessGroup taint-related condition.** We will add two [ProcessGroupConditionType](https://github.com/FoundationDB/fdb-kubernetes-operator/blob/main/api/v1beta2/foundationdbcluster_types.go#L673):

- `NodeTaintDetected` represents a ProcessGroup Node is tainted, but not long enough to replace it.
- `NodeTaintReplacing` represents a ProcessGroup whose Node has been tainted for longer-than-configured time and the ProcessGroup should be replaced.
```

Same as above in this context we are talking about `ProcessGroup`s and not `Pod`s. I personally would prefer a listing instead of a long sentence.